pkg/cloud/services/securitygroup/securitygroups.go

```diff
@@ -868,3 +942,26 @@ func (s *Service) getIngressRuleToAllowVPCCidrInTheAPIServer() infrav1.IngressRu
 		},
 	}
 }
+
+func generateIngressRuleKey(rule infrav1.IngressRule) string {
+	return fmt.Sprintf("%v-%v-%v", rule.Protocol, rule.FromPort, rule.ToPort)
+}
+
+func validateForDuplicateRule(existing []string, ruleKey string) bool {
+	for _, key := range existing {
+		if key == ruleKey {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
You can use the [stdlib package `slices`](https://pkg.go.dev/slices) instead of writing your own function.

Reply:
Sadly, you can't. Go mod file specifies 1.20 and so does the docker build. Slice is a 1.21 thing.

Reply:
Maybe `sets.New[string]` and using the `Has` function?